gtk4-macros/src/attribute_parser.rs

```diff
@@ -1,9 +1,81 @@
 // Take a look at the license at the top of the repository in the LICENSE file.
 
+use anyhow::{bail, Result};
 use proc_macro2::Span;
 use syn::parse::Error;
 use syn::spanned::Spanned;
-use syn::{Attribute, Field, Fields, Ident, Lit, Meta, MetaNameValue, NestedMeta, Type};
+use syn::{
+    Attribute, DeriveInput, Field, Fields, Ident, Lit, Meta, MetaList, MetaNameValue, NestedMeta,
+    Type,
+};
+
+pub enum TemplateSource {
+    File(String),
+    Resource(String),
+    String(String),
+}
+
+pub fn parse_template_source(input: &DeriveInput) -> Result<TemplateSource> {
+    let meta = match find_attribute_meta(&input.attrs, "template")? {
+        Some(meta) => meta,
+        _ => bail!("Missing 'template' attribute"),
+    };
+
+    let meta = match meta.nested.iter().find(|n| match n {
+        NestedMeta::Meta(m) => {
+            let p = m.path();
+            p.is_ident("file") || p.is_ident("resource") || p.is_ident("string")
+        }
+        _ => false,
+    }) {
+        Some(meta) => meta,
+        _ => bail!("Invalid meta, specify one of 'file', 'resource', or 'string'"),
+    };
+
+    let (ident, v) = parse_attribute(meta)?;
+
+    match ident.as_ref() {
+        "file" => Ok(TemplateSource::File(v)),
+        "resource" => Ok(TemplateSource::Resource(v)),
+        "string" => Ok(TemplateSource::String(v)),
+        s => bail!("Unknown enum meta {}", s),
+    }
+}
+
+// find the #[@attr_name] attribute in @attrs
```

Comment:
Why not proper rustdocs?

Reply:
To be totally honest these two helpers are cut and pasted from the glib macros (including the comments)... I guess the better question is: is there a place to put shared helper code?

Reply:
For now I left the cut&pasted versions, maybe we can do this refactoring when working on the gtk3 backport and then forward port here.

Reply:
> For now I left the cut&pasted versions, maybe we can do this refactoring when working on the gtk3 backport and then forward port here.

It could use an issue in gtk-rs/gtk-rs once the gtk3 PR is merged :)